### ISSUE TITLE
[Question]: Nomenclature integration

### ISSUE THREAD

**Issue by saona-raimundo** [tectonic-typesetting/tectonic#875]
I would like to use nomenclatures with tectonic.

I tried the simplest example from overleaf:
```tex
\documentclass{article}
\usepackage{nomencl}
\makenomenclature

\begin{document}
Here is an example:
\nomenclature{\(c\)}{Speed of light in a vacuum}
\nomenclature{\(h\)}{Planck constant}

\printnomenclature
\end{document}
```

The result was compiling two files: `.pdf` and `.nlo`, but the pdf does not contain the nomenclature.

What should I do to make it work? 
Is this a problem with the extra compilation step that the `nomencl` package needs?

**Comment by ratmice** [comment]
Yeah, this is because of the extra `makeindex` step needed, If I put the example in foo.tex
```
tectonic foo.tex
makeindex foo.nlo -s nomencl.ist -o foo.nls
tectonic foo.tex
```

the nomenclature appears in the document.

**Comment by saona-raimundo** [comment]
Perfect! Works like a charm!

Any chance this is automatically included in the tectonic compilation loop?
Tectonic already loops between TeX and BibTeX, so it may as well loop nomenclature, maybe if a nomenclature is encountered while reading the TeX code(?)

If it is a good idea to incorporate this step in the tectonic compilation loop, and thinking that the spirit of tectonic is to be reproducible and self-contained, should the external command `makeindex` be assimilated somehow? (I do not how complicated is this, I am just asking)

**Comment by ratmice** [comment]
I don't know, I have lately been using the [acro](https://ctan.org/pkg/acro) package instead, it doesn't work with tectonic yet, as it requires recent engine updates, but on engines it does work on doesn't require the `makeindex` step, I at least would be more likely to try and spend effort towards getting the `acro` package working but it isn't a part of the engine i'm at all familiar with.

**Comment by pkgw** [comment]
@ratmice do you know if acro would/should work when I finally update to TeXLive 2021/2022? Or are there other engine updates that need to happen?

I believe that it would be pretty easy to make `makeindex` a built-in tool similar to `bibtex` — I believe that it has a pretty simple WEB implementation that would be straightforward to ingest similar to `bibtex.c`. I've actually been kind of shocked that over the years I don't think anyone has requested `makeindex` support, which is certainly a big reason that I haven't looked into it more.

**Comment by ratmice** [comment]
@pkgw I believe just updating to 2021 should cover it, IIRC it depends on some new stuff in `expl3`, and those in turn required engine updates, alas I'm struggling to recall the specific details/get a specific useful minimal example of what needs work.

**Comment by pkgw** [comment]
Thanks! I'm going to close this issue since there's a solution for the initial question, but feel free to keep commenting or file a related issue if there's more to discuss here.

**Comment by saona-raimundo** [comment]
> I believe that it would be pretty easy to make makeindex a built-in tool similar to bibtex — I believe that it has a pretty simple WEB implementation that would be straightforward to ingest similar to bibtex.c. I've actually been kind of shocked that over the years I don't think anyone has requested makeindex support, which is certainly a big reason that I haven't looked into it more.

Looking at [some `makeindex` information](https://latexref.xyz/makeindex.html), at the very end it is mentioned that 
""
A more recent indexing program supporting Unicode is xindex, written in Lua (https://ctan.org/pkg/xindex). 
""

Both [`xindy`](https://ctan.org/pkg/xindy) and [`xindex`](https://ctan.org/pkg/xindex) seem to be product of adding i18n to `makeindex` (in particular unicode support). So, although digesting `makeindex` into tectonic might be possible, it seems it would be a temporary solution if tectonic has the goal of unicode support (just as bibtex needs to be replaced by something like biblatex). Am I understanding well the goal of tectonic about unicode support, right?

**Comment by pkgw** [comment]
Yes, you are absolutely correct about Unicode support. I *think* that classic makeindex might be simple enough that making it work with UTF-8 might be pretty straightforward, or at least more straightforward than trying to port xindex from Lua. But I haven't looked into that.

**Comment by ratmice** [comment]
Just wanted to mention that I just checked the acro package against the Tectonic TeXLive 2021 update.

So while this doesn't get us the nomenclature package working out of the box, it is an alternative package people can use instead perhaps (once everything lands).

**Comment by mnrvwl** [comment]
Today, while working with some sources I remembered of this https://github.com/tectonic-typesetting/tectonic/issues/875, and https://github.com/tectonic-typesetting/tectonic/issues/115.

In the sources, there a compilation step like the one mentioned above:
```console
latex test.tex
makeindex test.nlo -s nomencl.ist -o test.nls
latex test.tex
```

Automated tools like `latexmk` and `tectonic` are incredibly convenient when developing documents, allowing us to avoid manually maintaining the required compilation steps and save time figuring out which steps need to be called for which packages.

With `latexmk`, we can simplify the compilation process with a single, simple command call, but building the index will require something more, using a `.latexmkrc` file —stored per repo or in the user's XDG_CONFIG— or the `-e` flag:
```console
latexmk -e 'add_cus_dep("nlo", "nls", 0, "nlo2nls");                                 
sub nlo2nls {
    system("makeindex $_[0].nlo -s nomencl.ist -o $_[0].nls -t $_[0].nlg");
}' test.tex
```

Which is fairly easy to find, and you can add it to a `.latexmkrc`, that works per repo or in users XDG_CONFIG. We can also pass it via the `-e` flag, i.e.: `latexmk -e 'add_cus_dep( [...]`.

Tectonic doesn't support this, and it means every package that needs something like that needs a workaround of multiple steps, losing our nice, cozy, automated wrapping. 

IIRC is possible to do it directly in the source with `imakeidx` and `xindy`, even if sometimes shell-escaping applies, and we have shell-escape in Tectonic:
```latex
\usepackage{imakeidx}
  \makeindex[options=-s nomencl.ist]
```

However, `nomencl` uses `makeidx` under the hood, preventing us from solving it through a config directly in the code. Other packages that rely on more recent takes on the same thing —indexers— should be okay or workable in `tectonic`. Both `acro`, `glossaries` are alternative solutions, and I'd like to know whether `nomencl` may still be preferred in certain use cases.

In the TeX ecosystem, many packages require special treatment, and some platforms have [good examples](https://www.overleaf.com/learn/how-to/How_does_Overleaf_compile_my_project%3F#A_listing_of_Overleaf%E2%80%99s_LatexMk_file_(2022)) of workarounds they need.

Overall, I think it would be interesting to discuss the topic of supporting fine-grained control for special packages in tectonic, or decide to encourage users to a known set of alternatives.

@pkgw Feel free to take my comment and open a discussion with it.

**Comment by ratmice** [comment]
> Both `acro`, `glossaries` are alternative solutions, and I'd like to know whether `nomencl` may still be preferred in certain use cases.

FWIW, I've found that `acro` fits my needs much better than the `nomencl` package ever did, so I would personally never want to go back.
It's been a while since I've used nomencl.  Hopefully I remember correctly but with `acro` I have been able to make a single file containing lists of acronyms which are shared across many documents.  It contains many more acronyms than are used in a single document (usually while I have 4 or 5 lists of acronyms I only actually pull a few acronyms from 1 or 2 of those lists).
via https://github.com/cgnieder/acro/issues/228

Using the custom templates, I can just include *all* the lists of acronyms in the same way in each document,
It'll only print the acronyms & lists of acronyms which have entries that are used.  So exactly the output I want without having to customize a single thing for any individual document.  So not only does it not require special build instructions, but also has given me the exact functionality I want from it.
